### DISCUSSION TITLE
Entities not discovered

### DISCUSSION THREAD

**Discussion by yackinn** [mikro-orm/mikro-orm#2864] [Q&A] [answered]
**Describe the bug**

**Stack trace**
```
[Nest] 16845  - 05.03.2022, 19:17:22   ERROR [ExceptionHandler] No entities were discovered
MetadataError: No entities were discovered
    at Function.noEntityDiscovered (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@mikro-orm/core/errors.js:151:16)
    at MetadataValidator.validateDiscovered (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@mikro-orm/core/metadata/MetadataValidator.js:38:42)
    at MetadataDiscovery.findEntities (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:91:24)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at MetadataDiscovery.discover (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:34:9)
    at MikroORM.discoverEntities (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@mikro-orm/core/MikroORM.js:90:25)
    at Function.init (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@mikro-orm/core/MikroORM.js:45:9)
    at Injector.instantiateClass (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@nestjs/core/injector/injector.js:333:37)
    at callback (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@nestjs/core/injector/injector.js:48:30)
    at Injector.resolveConstructorParams (/Users/yannick.boetzkes/Development/playground/tutorials/nestjs/udemy-nestjs-the-complete-developers-guide/car-value-api/node_modules/@nestjs/core/injector/injector.js:122:24)
```

**To Reproduce**
Generated app via nest cli with mikro orm added as dependency.
The entity is present in the src and dist folder as it's found using the exact path.

// Error
MikroOrmModule.forRoot({
      host: "localhost",
      port: 3306,
      password: "password",
      entities: ['./dist/\*\*/entities'],
      entitiesTs: ['./src/**/entities'],

// Works
MikroOrmModule.forRoot({
      host: "localhost",
      port: 3306,
      password: "password",
      entities: ['./dist/user/entities'],
      entitiesTs: ['./src/user/entities'],

**Expected behavior**
Entities should be found using glob path.

**Additional context**

**Versions**

| Dependency | Version |
| - | - |
| node | 16.14.0 |
| typescript | 4.3.5 |
| mikro-orm | 5.0.5 |

**Comment by yackinn** [comment, marked as answer]
solved it by providing the mikroorm config in the package json

**Reply by productdevbook** [reply]
how to fixed ? 

example package.json share please

**Reply by biscuitdelicious** [reply]
Use `autoLoadEntities: true` in the root config. It works with [NestJS](https://mikro-orm.io/docs/usage-with-nestjs#load-entities-automatically)

```
MikroOrmModule.forRoot({
...
// entities: ['./dist/user/entities'],
entities: [User], // The one you created in the 
autoLoadEntities: true,
...
```

**Comment by yackinn** [comment]
This should be in the root level of your package.json.
```json
"mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      <path to your config>
    ]
  }
```

That's what we used.
```ts
...
let config: MikroOrmModuleSyncOptions = {
  type: 'postgresql',
  host: process.env.DB_HOST,
  port: Number(process.env.DB_PORT),
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  dbName: process.env.DB_NAME,
  debug: Boolean(process.env.DB_DEBUG),
  entities: [
    Booking,
    ...
  ],
  discovery: {
    disableDynamicFileAccess: true
  },
  loadStrategy: LoadStrategy.JOINED,
  populateWhere: PopulateHint.INFER,
  highlighter: new SqlHighlighter(),
  tsNode: true,
  migrations: {
    path: __dirname + '/migrations',
    pathTs: __dirname + '/migrations',
    disableForeignKeys: false
  },
  allowGlobalContext: true,
  logger: msg => logger.log(msg)
};

if (process.env.NODE_ENV === 'production') {
  config = {
    ...config,
    driverOptions: {
      connection: { ssl: { rejectUnauthorized: false } },
    }
  };
}

export default config;
```

**Reply by productdevbook** [reply]
```json
  "mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./src/mikro-orm.config.ts",
      "./dist/src/mikro-orm.config.js"
    ]
  }
```
  
  and some config but dont fixed

**Reply by alexrequelme** [reply]
Does it work for you if you build it and pack it with docker? Because it doesn't recognize me if the path is something like "./src/database/mikro-orm.config.js"

**Comment by manusanchev** [comment]
same problem here, I need to specify the path to my entities within a folder whose name changes throughout my project but it does not work

```
entities: ["./build/src/modules/**/entities"],
entitiesTs: ["./src/modules/**/entities"]
```

**Comment by manusanchev** [comment]
i fixed it by adding `/* ` to the end of path

```
entities: ["./build/src/modules/**/entities/*"],
entitiesTs: ["./src/modules/**/entities/*"]
```

@productdevbook maybe is your problem too

this is somewhat contradictory to what is stated in the docs

**Reply by productdevbook** [reply]
Thank you, yes fixed.

**Reply by B4nan** [reply]
Interesting, first time hearing this. Maybe its an issue with using `**` in the paths, as we have tests that take just a directory in there (but without using `**` in the path).

Will take a look, this should be easily fixable if it really behaves like that.

**Reply by AndrianTsyhulov** [reply]
This one actually helps a lot

I was running my stuff from Dockerfile and had my entities specified like that:

  ```
 entities: ['./dist/**/*.entity.js'],
   entitiesTs: ['./src/**/*.entity.ts'],
```

So getting rid of **dist**  folder in path fixed the issue

 ```
 entities: ['./**/*.entity.*'],
  entitiesTs: ['./**/*.entity.*'],
 
```

**Reply by B4nan** [reply]
> entities: ['./**/*.entity.*'],
> entitiesTs: ['./**/*.entity.*'],

This is wrong, those paths shouldn't be the same. Keep at least the file extensions. Maybe you need to adjust `baseDir`, or maybe you don't have TS support set up correctly.

**Comment by yackinn** [comment]
What command are you using to run the mikroorm cli?
Use this in your as a package.json script to run mikroorm in your terminal.

`"mikro-orm": "MIKRO_ORM_CLI_TS_CONFIG_PATH=apps/backend/tsconfig.app.json npx mikro-orm"`

`yarn mikro-orm ...`

**Reply by B4nan** [reply]
FYI you dont need to use `npx` here as yarn (as well as NPM) will add the NPM scripts to the path automatically for the package.json scripts. 

Also you can set the path to tsconfig via package.json settings, but I guess you want this for a monorepo setup and have multiple of them?

**Reply by yackinn** [reply]
Correct. Thank you.
Yes, we didn't want to make the config global for the entire monorepo. It should be tied to one application.

**Comment by RichardWright** [comment]
Can this be used with types? no urls specified?

**Reply by B4nan** [reply]
what do you mean by types and urls?

**Comment by RichardWright** [comment]
I did. Total red herring, database connected fine but due to a missing
index none of the initial queries functioned correctly

On Sat, 24 Sept 2022 at 00:07, Alexander Requelme ***@***.***>
wrote:

> did you manage to solve it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mikro-orm/mikro-orm/discussions/2864#discussioncomment-3720769>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGXJF7X5S3DSKRP3F73FM3V7YZZRANCNFSM5QAHTPWQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Thanks

Richard

**Comment by jacobsalway** [comment]
If anyone has issues with their entities not being discovered even with all the config set up correctly (which you can verify using `mikro-orm debug`), I did a bit of source code diving and discovered the metadata discovery skips files named `index.ts` or `index.js`. 

So if you're like me and had your entities defined under an `index.ts` file (`src/entities/index.ts` for me), try using a different file name.

**Source code permalinked below:**
https://github.com/mikro-orm/mikro-orm/blob/51a9ab67a40b85bc9e6ad031f8b277e38174eaa9/packages/core/src/metadata/MetadataDiscovery.ts#L261

**Comment by maciek-wu** [comment]
If anyone uses NestJS v10 as monorepo with pnpm workspace and is facing "No entities were discovered" issue despite correct configuration (mikro-orm debug all in green). Here are some tips:

- Make sure you have all packages updated to the latest versions (mikro-orm, typescript, ts-node, reflect-metadata etc.)
- Make sure that mikro-orm.config.ts is in the root of your project or you have `mikro-orm: configPaths` configured in the root package.json
- In Mikro-Orm config use only `entities` param (for some reason `entitiesTs` breaks entities discovery)
- Make sure you have added `node-linker='hoisted'` in root .npmrc and you have run `pnpm install --force` to rebuild node_packages
- Make sure you have app file structure in ./apps/your-app/dist folder with all entity JS files (not only a single main.js bundled by webpack)

@B4nan FYI, I was stuck because of the `entitiesTs` param was breaking config for some reason. Maybe it's because of some changes in the NestJS v10, any idea? BTW. great job with this ORM, keep it up! Cheers :)

**Reply by B4nan** [reply]
Hard to guess without a complete reproduction, having a working example repo for this would be very welcome.

**Reply by B4nan** [reply]
Btw for more complex setups with tsmorph (which I believe you use, otherwise it shouldn't be really important to use `entitiesTs` option) you could also try the generated cache adapter and building the metadata via CLI instead.

**Reply by moforemmanuel** [reply]
Had same issue, `entitiesTs` sure was the culprit. When I made it `undefined` in production, worked like a charm. Sample config below:

```
import { defineConfig, PostgreSqlDriver } from '@mikro-orm/postgresql';
import dotenv from 'dotenv';
dotenv.config();

export default defineConfig({
  entities: ['./dist/**/*.entity.js'],
  entitiesTs:
    process.env.NODE_ENV === 'development'
      ? ['./src/**/*.entity.ts']
      : undefined,
  migrations: {
    tableName: 'mikroorm_migrations',
    path: 'dist/database/migrations',
    pathTs: 'src/database/migrations',
    glob: '!(*.d).{js,ts}',
    transactional: true,
    allOrNothing: true,
    emit: 'ts',
  },
  driver: PostgreSqlDriver,
  clientUrl: process.env.DATABASE_URL,
  pool: {
    max: parseInt(process.env.DATABASE_POOL_MAX as string, 10) || 20,
  },
  driverOptions: {
    connection: {
      ssl:
        process.env.NODE_ENV !== 'development' &&
        process.env.AIVEN_CA_CERT_BASE64
          ? {
              rejectUnauthorized: true,
              ca: Buffer.from(
                process.env.AIVEN_CA_CERT_BASE64,
                'base64',
              ).toString('utf8'),
            }
          : undefined,
    },
  },
  debug: process.env.NODE_ENV === 'development',
});

```

NestJS v11 btw!

Thank you @maciek-wu

**Comment by biscuitdelicious** [comment]
A bit late but why not use `autoLoadEntities: true` in the root config? It works with [NestJS](https://mikro-orm.io/docs/usage-with-nestjs#load-entities-automatically)

Instead of:

```
MikroOrmModule.forRoot({
host: "localhost",
port: 3306,
password: "password",
entities: ['./dist/user/entities'],
entitiesTs: ['./src/user/entities'],
```

use

```
MikroOrmModule.forRoot({
host: "localhost",
port: 3306,
password: "password",
entities: [Users], // Given as example, the entity you have created in your folder
autoLoadEntities: true, // <- 
```
